### PR TITLE
Add caching support for model testing

### PR DESCRIPTION
**Pull request by saienduri**
This commit adds caching support for model testing. This brings the download time from ~20 mins to 23 seconds. With these changes, it is feasible to add model testing as a presubmit in the iree repo which was the reason for these changes. 

Description of what's going on:  
First, we use an environment variable passed in (IREE_TEST_FILES) to determine if we even want to use a cache or not (don't want to make it a requirement). In CI, we set this because we want a persistent runner cache. If this is set, we download files to the cache location specified and symlink it to the testing directory. Otherwise, we just directly download into the repo, and it won't be using any caching mechanism. The test runner doesn't need to know about caching at all, and even initially, it will just create the cache and download there if it doesn't exist. There is no manual interaction that needs to take place for any system or person that wants to use this.

Example of actions run first time and second time (runner knows nothing about caching, no manual setup):
https://github.com/nod-ai/SHARK-TestSuite/actions/runs/8711532825
https://github.com/nod-ai/SHARK-TestSuite/actions/runs/8711798501

Final model testing job time: ~5 mins

The machine in the iree repo is even faster, so can expect lower times :)

Fixes https://github.com/nod-ai/SHARK-TestSuite/issues/169

### CODE REVIEW

File: iree_tests/download_remote_files.py [hunk 1]
```diff
@@ -81,14 +83,26 @@ def download_azure_remote_file(test_dir: Path, remote_file: str):
         blob_size_str = human_readable_size(blob_properties.size)
         remote_md5 = get_remote_md5(remote_file, blob_properties)
 
-        local_file_path = test_dir / remote_file_name
+        cache_location = os.getenv("IREE_TEST_FILES", default="")
+        if cache_location == "":
+            os.environ["IREE_TEST_FILES"] = str(REPO_ROOT)
+            cache_location = REPO_ROOT
+        if cache_location == REPO_ROOT:
```
Comment (ScottTodd) [inline iree_tests/download_remote_files.py:90]: String and Path handling here could be tricky. We might want to convert the string coming from the environment variable to a canonical path before handling it further. 

Could use https://docs.python.org/3/library/pathlib.html#pathlib.Path.resolve for that

File: iree_tests/download_remote_files.py [hunk 2]
```diff
@@ -81,14 +83,26 @@ def download_azure_remote_file(test_dir: Path, remote_file: str):
...
+            local_dir_path = test_dir
+            local_file_path = test_dir / remote_file_name
+        else:
+            cache_location = os.path.expanduser(cache_location)
```
Comment (ScottTodd) [inline iree_tests/download_remote_files.py:94]: Related to my other comment, I think `resolve()` covers more cases than just `expanduser`

File: iree_tests/download_remote_files.py [hunk 3]
```diff
@@ -81,14 +83,26 @@ def download_azure_remote_file(test_dir: Path, remote_file: str):
...
+            local_file_path = Path(cache_location) / "iree_tests" / relative_dir / remote_file_name
+
         local_md5 = get_local_md5(local_file_path)
 
         if remote_md5 and remote_md5 == local_md5:
             logger.info(
                 f"  Skipping '{remote_file_name}' download ({blob_size_str}) "
                 "- local MD5 hash matches"
             )
+            os.symlink(local_file_path, test_dir / remote_file_name)
```
Comment (ScottTodd) [inline iree_tests/download_remote_files.py:105]: Let's add some logging for the symlink setup too. (We want enough information in the logs for future debugging)

File: iree_tests/download_remote_files.py [hunk 4]
```diff
@@ -102,9 +116,13 @@ def download_azure_remote_file(test_dir: Path, remote_file: str):
                 f"to '{relative_dir}' (local MD5 does not match)"
             )
 
+        if not os.path.isdir(local_dir_path):
+            os.makedirs(local_dir_path)
         with open(local_file_path, mode="wb") as local_blob:
             download_stream = blob_client.download_blob(max_concurrency=4)
             local_blob.write(download_stream.readall())
+        if str(cache_location) != str(REPO_ROOT):
+            os.symlink(local_file_path, test_dir / remote_file_name)
```
Comment (ScottTodd) [inline iree_tests/download_remote_files.py:125]: Could add logging here

File: iree_tests/download_remote_files.py
Comment (ScottTodd) [inline iree_tests/download_remote_files.py:1]: I'm wondering if we should split the scripts into some separate scripts / stages so they can be used piecemeal, like `download_remote_files.py`, `populate_cache.py`, `configure_cache.py`. That can be done in a follow-up.

Steady state CI jobs (with cache):
* `download_remote_files.py` or `populate_cache.py` (downloads into the cache)
* `configure_cache.py` (sets up symlinks)

CI maintenance (e.g. first time adding a runner):
* `download_remote_files.py` or `populate_cache.py` (downloads into the cache)

Local runs:
* `download_remote_files.py`

Basically trying to avoid overloading a single script with environment variables and flags -- it should be obvious which script(s) to run in each scenario.

**Review by ScottTodd — APPROVED** [review]
Cool! This looks great to get us going.

File: iree_tests/download_remote_files.py [hunk 5]
```diff
@@ -103,6 +103,9 @@ def download_azure_remote_file(test_dir: Path, remote_file: str):
                 "- local MD5 hash matches"
             )
             os.symlink(local_file_path, test_dir / remote_file_name)
+            logger.info(
+                f"  Created symlink for '{local_file_path}' to '{test_dir / remote_file_name}' "
+            )
```
Comment (ScottTodd) [inline iree_tests/download_remote_files.py:108]: nit: can drop trailing ` ` (space). I included those in other sections since the string continued on the next line

File: iree_tests/download_remote_files.py [hunk 6]
```diff
@@ -123,6 +126,9 @@ def download_azure_remote_file(test_dir: Path, remote_file: str):
             local_blob.write(download_stream.readall())
         if str(cache_location) != str(REPO_ROOT):
             os.symlink(local_file_path, test_dir / remote_file_name)
+            logger.info(
+                f"  Created symlink for '{local_file_path}' to '{test_dir / remote_file_name}' "
+            )
```
Comment (ScottTodd) [inline iree_tests/download_remote_files.py:131]: same here

**Review by ScottTodd — APPROVED** [review]

File: iree_tests/download_remote_files.py
Comment (saienduri) [reply iree_tests/download_remote_files.py:1]: Yeah, I can see the value in this follow-up instead of having one beast of a file. Makes it easier to understand what is going on and reuse utilities implemented in other places.

File: iree_tests/download_remote_files.py (hunk 5 above)
Comment (saienduri) [reply iree_tests/download_remote_files.py:108]: Ah I see. I thought it was some sort of styling format for logging, thanks
